### ISSUE TITLE
Hard to follow crate_universe documentation

### ISSUE THREAD

**Issue by jfirebaugh** [bazelbuild/rules_rust#1674]
I'm trying to follow this example in https://bazelbuild.github.io/rules_rust/crate_universe.html#direct-packages:

```bzl
load("@cargo_bazel//:defs.bzl", "crate", "crates_repository", "render_config")

crates_repository(
    name = "crate_index",
    lockfile = "//:Cargo.Bazel.lock",
    packages = {
        "async-trait": crate.spec(
            version = "0.1.51",
        ),
        "mockall": crate.spec(
            version = "0.10.2",
        ),
        "tokio": crate.spec(
            version = "1.12.0",
        ),
    },
    # Setting the default package name to `""` forces the use of the macros defined in this repository
    # to always use the root package when looking for dependencies or aliases. This should be considered
    # optional as the repository also exposes alises for easy access to all dependencies.
    render_config = render_config(
        default_package_name = ""
    ),
)

load("@crate_index//:defs.bzl", "crate_repositories")

crate_repositories()
```

This produces the following error:

```
ERROR: Failed to load Starlark extension '@cargo_bazel//:defs.bzl'.
Cycle in the workspace file detected. This indicates that a repository is used prior to being defined.
The following chain of repository dependencies lead to the missing definition.
 - @cargo_bazel
This could either mean you have to add the '@cargo_bazel' repository with a statement like `http_archive` in your WORKSPACE file (note that transitive dependencies are not added automatically), or move an existing definition earlier in your WORKSPACE file.
ERROR: cycles detected during target parsing
```

If I replace `@cargo_bazel//:defs.bzl` with `@rules_rust//crate_universe:defs.bzl`, then:

```
ERROR: /Users/john/figma/rules_rust_built/WORKSPACE:21:18: //external:crate_index: missing value for mandatory attribute 'cargo_lockfile' in 'crates_repository' rule
ERROR: error loading package '': Encountered error while reading extension file 'defs.bzl': no such package '@crate_index//': error loading package 'external': Could not load //external package
```

If I replace `lockfile` with `cargo_lockfile`, then:

```
ERROR: An error occurred during the fetch of repository 'crate_index':
   Traceback (most recent call last):
	File "/private/var/tmp/_bazel_john/6ebb6fa28311e23a56b6491e5d493861/external/rules_rust/crate_universe/private/crates_repository.bzl", line 34, column 30, in _crates_repository_impl
		lockfiles = get_lockfiles(repository_ctx)
	File "/private/var/tmp/_bazel_john/6ebb6fa28311e23a56b6491e5d493861/external/rules_rust/crate_universe/private/generate_utils.bzl", line 299, column 36, in get_lockfiles
		cargo = repository_ctx.path(repository_ctx.attr.cargo_lockfile),
Error in path: Not a regular file: /Users/john/figma/rules_rust_built/Cargo.Bazel.lock
```

Okay, I read further down in the documentation, it says "If set, this file must exist within the workspace (but can be empty) before this rule will work." That's referring to the `lockfile` attribute, not `cargo_lockfile`. I can maybe guess that it also applies to `cargo_lockfile`, but it would be nice if the error message said something like "Create an empty Cargo.Bazel.lock file in order to proceed" right away, instead of "Error in path: Not a regular file". Or, why can't it just create the file for me?

**Comment by bazaglia** [comment]
I agree that the documentation for that isn't the best. You need to provide both `cargo_lockfile` and `lockfile`. For generating your `lockfile`, that will be holding some json content, run `CARGO_BAZEL_REPIN=1 bazel sync --only=crate_index` or any other command prefixed with `CARGO_BAZEL_REPIN=1`.

**Comment by illicitonion** [comment]
Thanks for the bug report - I definitely agree with all of the points you've raised!

On the `@cargo_bazel` reference and the `lockfile`/`cargo_lockfile` confusion, I've just raised https://github.com/bazelbuild/rules_rust/pull/1677

I also agree with the "Error in path" issue, but sadly this one is much harder for us to fix because of some Bazel limitations. Fortunately, we should be able to give a much better experience from Bazel 6, so I've filed https://github.com/bazelbuild/rules_rust/issues/1676 for us to improve that when Bazel 6 is released.

We're very open to more feedback as to how we can get the first-run experience as smooth as possible!

**Comment by jfirebaugh** [comment]
Thanks you for the documentation improvements!

I'm still confused by the `lockfile`/`cargo_lockfile` distinction in the "Cargo free" setup (the example I was trying to follow). You changed it to:

https://github.com/bazelbuild/rules_rust/blob/605f427bf7c016f585cf815bc0f38175859bf76e/crate_universe/docs.bzl#L107-L108

But in a Cargo-free setup, `Cargo.lock` won't exist? Is the expectation that you create both of these as empty files and rules_rust generates their content, or what?

**Comment by illicitonion** [comment]
I believe currently both are required, but that repinning should generate both if they're empty; I think in cargo-free mode we could probably make the `cargo_lockfile` optional and either derive the information either from your `packages` attribute or sneak the information into the other `lockfile`, but that would require a bit of design work.

Conceptually, the `cargo_lockfile` is information used when repinning ("If I'm repinning and just want to keep my existing dependencies as they were before and pull in new versions of anything I didn't have before, here are the versions I should preserve"), and `cargo_lockfile is information used to avoid needing to repin ("If nothing has changed, I can use these pre-computed values from last time"), so they have slightly different purposes.

Alternatively, I can imagine making `cargo_lockfile` optional, and in that case we'd simply stop supporting conservative repinning (i.e. every repin would update all dependencies) if you don't have a `cargo_lockfile`.
